conn.go

```diff
@@ -69,13 +69,19 @@ type Conn struct {
 	writeHeaderBuf [8]byte
 	writeHeader    header
 
+	// Close handshake state.
+	closeStateMu     sync.RWMutex
+	closeReceivedErr error
+	closeSentErr     error
+
+	// CloseRead state.
 	closeReadMu   sync.Mutex
 	closeReadCtx  context.Context
 	closeReadDone chan struct{}
 
-	closed  chan struct{}
+	closing atomic.Bool
 	closeMu sync.Mutex
-	closing bool
+	closed  chan struct{}
```

Comment:
It's not immediately clear what `closeMu` is protecting here. I think it would be helpful to make this more explicit.

Ditto for `closing`, we could be more clear about what its purpose is here for future readers.

Reply:
I can add a comment as it only protects the `closed` channel now. Previously it protected the `closing` boolean as well, but I changed it to atomic to avoid a deadlock in rare situations. For example when `msgWriter` is being closed at the same time as reading a close frame, resulting in writing a close frame as response. The mutexes between conn close and msg writers are somewhat intertwined and I'm thinking of doing a larger refactor to address this at some point.